homeassistant/components/octoprint/octoprintapi.py

```diff
@@ -0,0 +1,143 @@
+"""Simple JSON wrapper for OctoPrint's API."""
+import logging
+import time
+
+from aiohttp.hdrs import CONTENT_TYPE
+import requests
+
+from homeassistant.const import CONTENT_TYPE_JSON
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class OctoPrintAPI:
+    """Simple JSON wrapper for OctoPrint's API."""
+
+    def __init__(self, api_url, key, bed, number_of_tools):
+        """Initialize OctoPrint API and set headers needed later."""
+        self.api_url = api_url
+        self.headers = {CONTENT_TYPE: CONTENT_TYPE_JSON, "X-Api-Key": key}
+        self.printer_last_reading = [{}, None]
+        self.job_last_reading = [{}, None]
+        self.job_available = False
+        self.printer_available = False
+        self.printer_error_logged = False
+        self.available = False
+        self.available_error_logged = False
+        self.job_error_logged = False
+        self.bed = bed
+        self.number_of_tools = number_of_tools
+
+    def get_tools(self):
+        """Get the list of tools that temperature is monitored on."""
+        tools = []
+        if self.number_of_tools > 0:
+            for tool_number in range(0, self.number_of_tools):
+                tools.append(f"tool{tool_number!s}")
+        if self.bed:
+            tools.append("bed")
+        if not self.bed and self.number_of_tools == 0:
+            temps = self.printer_last_reading[0].get("temperature")
+            if temps is not None:
+                tools = temps.keys()
+        return tools
+
+    def get(self, endpoint):
+        """Send a get request, and return the response as a dict."""
+        # Only query the API at most every 30 seconds
+        now = time.time()
+        if endpoint == "job":
+            last_time = self.job_last_reading[1]
+            if last_time is not None:
+                if now - last_time < 30.0:
+                    return self.job_last_reading[0]
+        elif endpoint == "printer":
+            last_time = self.printer_last_reading[1]
+            if last_time is not None:
+                if now - last_time < 30.0:
+                    return self.printer_last_reading[0]
+
+        url = self.api_url + endpoint
+        try:
+            response = requests.get(url, headers=self.headers, timeout=9)
```

Comment:
We need to replace the direct request calls with a library before we add config flow.

Please extract the device specific interface code and data parsing to a standalone library published on PyPI.

https://developers.home-assistant.io/docs/creating_component_code_review#4-communication-with-devicesservices

Reply:
Was planning on doing that as the next phase of updating the integration. I am currently working on the lib, but I found swapping to config_flow then introducing the new library/DataCoordinator usages to be a cleaner PR.  Given this is a direct lift from the existing __init__ with no changes, can we proceed with the lib happening next?

Reply:
If we use a separate target branch than dev branch to allow multiple PRs before releasing anything, I'd be ok with it.

We generally don't allow major changes to integrations that don't have a library.

What do you think @frenck?

Reply:
I think this should be done the other way around :) But either way is fine 👍

Reply:
@rfleming71 let me know if I should push a copy of dev branch upstream that you can rebase onto, or if you want to start with the library instead.

Reply:
Would the preference be to point this PR is a second branch and have the lib follow onto that branch or get the lib merged into dev first then rebase this on top? May be a few days before the lib PR is ready.

Reply:
My preference would be the copy of the dev branch, but I will do which ever process is typically followed. May be a day or two before the lib PR is ready

Reply:
Here's a fresh copy of dev branch:
https://github.com/home-assistant/core/tree/rework-octoprint

Please rebase your branch on that and change target branch here in the PR.

Reply:
Done. Thanks!